---
output: 
  readmedown::plain_document:
    columns: 80
bibliography: input_data_readme_refs.yaml
---

```{r manual, include = FALSE, evaluate = FALSE}
# Run these lines if rendering outside of targets plan
library(targets)
library(tarchetypes)
library(readmedown)
source(here::here("R/packages.R"))
source(here::here("R/functions.R"))
```

```{r setup, include = FALSE, cache = FALSE}
knitr::opts_chunk$set(
  echo = FALSE, message = FALSE,
  warning = FALSE, results = "hide", cache = FALSE)

# Load objects from targets workflow (all paths to data files)
tar_load(
  c(
    date_cutoff,
    gb_release,
    gb_readme_path,
    restez_sql_db_archive,
    ref_aln_archive,
    ref_aln_files,
    taxdump_zip_file)
  )
```

This README.txt file was generated on `r Sys.Date()` by Joel Nitta

-------------------- 

GENERAL INFORMATION 

------------------

Title of Dataset: Fern Tree of Life (FTOL) input data

Principal Investigator: Joel H. Nitta

Department of Integrated Biosciences, Graduate School of Frontier Sciences,
The University of Tokyo, Chiba, Japan. joelnitta@gmail.com

Associate or Co-investigators: Eric Schuettpelz, Santiago Ramírez-Barahona, Wataru Iwasaki

Date of data collection: 1990 - `r lubridate::year(lubridate::ymd(date_cutoff))`

Geographic location of data collection: Global

Information about funding sources that supported the collection of the data:
Funding provided in part by the Japan Society for the Promotion of Science
(Kakenhi) Grant Numbers 16H06279, 22H04925, and 22K15171 and the Smithsonian
National Museum of Natural History Peter Buck Fellowship (JHN).

--------------------------

SHARING/ACCESS INFORMATION

--------------------------

Licenses/restrictions placed on the data: CC0 v1.0 license

Links to publications that cite or use the data:

Nitta JH, Schuettpelz E, Ramírez-Barahona S, Iwasaki W. 2022. An open and 
continuously updated fern tree of life. Frontiers in Plant Sciences 13
https://doi.org/10.3389/fpls.2022.909768.

Links to other publicly accessible locations of the data: none.

Links/relationships to ancillary data sets:

- ferncal (https://github.com/fernphy/ferncal)
- pteridocat (https://github.com/fernphy/pteridocat)

Was data derived from another source? Yes, in part from GenBank (https://www.ncbi.nlm.nih.gov/genbank/),
which places no restrictions on its use or distribution.

Recommended citation for this dataset (click FigShare "Cite" button for specific DOI version): 

FTOL Working Group (`r lubridate::year(Sys.Date())`). Fern Tree of Life (FTOL) input data.
https://doi.org/10.6084/m9.figshare.19474316

--------------------

DATA & FILE OVERVIEW

--------------------

File List:

- README.genbank: Text file including information about GenBank release `r gb_release`.
- restez_sql_db.tar.gz: Compressed file archive including database of fern DNA sequences from GenBank release `r gb_release`.
- ref_aln.tar.gz: Compressed file archive including one fasta file per target plastid locus.
- taxdmp.zip: Compressed file archive including NCBI taxonomic data.

--------------------------

METHODOLOGICAL INFORMATION

--------------------------

The data included here are used in a pipeline that (mostly) automatically
generates a maximally sampled fern phylogenetic tree based on plastid sequences
in GenBank (https://github.com/fernphy/ftol).

The first step is to download the latest release of GenBank data from the NCBI
GenBank FTP site (https://ftp.ncbi.nlm.nih.gov/genbank/) and use it to create a
local database of fern sequences. This is done with custom R scripts contained
in https://github.com/fernphy/ftol, in particular setup_gb.R
(https://github.com/fernphy/ftol/blob/main/R/setup_gb.R).

Next, a set of reference FASTA files for 79 target loci (one per locus;
ref_aln.tar.gz) is generated. These include 77 protein-coding genes based on a
list of 83 genes [@Wei2017a] that was filtered to only genes that show no
evidence of duplication, plus two spacer regions (trnL-trnF and rps4-trnS). Each
FASTA file in ref_aln.tar.gz includes one representative (longest) sequence per
avaialable fern genus. This is done with prep_ref_seqs_plan.R
(https://github.com/fernphy/ftol/blob/main/prep_ref_seqs_plan.R).

Sequences matching the target loci are then extracted from each accession in the
local database using the FASTA files contained in ref_aln.tar.gz as references
with the "Reference_Blast_Extract.py" script of superCRUNCH [@Portik2020].

The extracted sequences are aligned with MAFFT [@Katoh2002], phylogenetic
analysis is done using IQ-TREE [@Nguyen2015] and divergence times estimated with
treePL [@Smith2012]. 

For additional methodological details, see:

Nitta JH, Schuettpelz E, Ramírez-Barahona S, Iwasaki W. 2022. An open and 
continuously updated fern tree of life. Frontiers in Plant Sciences 13
https://doi.org/10.3389/fpls.2022.909768.

--------------------------

DATA-SPECIFIC INFORMATION

--------------------------

`r path_file(gb_readme_path)`: Text file including information about GenBank
release `r gb_release` downloaded from the NCBI GenBank FTP site
(https://ftp.ncbi.nlm.nih.gov/genbank/).

MD5 checksum: `r tools::md5sum(gb_readme_path)`

--------------------------

`r path_file(restez_sql_db_archive)`: Compressed file archive including fern DNA
sequences downloaded from the NCBI GenBank FTP site (release `r gb_release`,
https://ftp.ncbi.nlm.nih.gov/genbank/) formatted as a duckDB database
(https://duckdb.org/), created with restez R package
v`r as.character(packageVersion("restez"))` (https://github.com/ropensci/restez).
The sequences can also be accessed with restez. To do so, the contents of the
archive ("sql_db" file) needs to be extracted to a folder "restez". Then, run
the restez function restez_path_set() with the folder containing "restez/sql_db"
as input.

MD5 checksum: `r tools::md5sum(restez_sql_db_archive)`

--------------------------

`r path_file(ref_aln_archive)`: Compressed file archive including one fasta file
per target platid locus (`r length(ref_aln_files)` loci total). Each FASTA file
is named by the locus. Sequences within each FASTA file named by species and
GenBank accession number separated by an underscore. One representative
(longest) sequence per fern genus is included per locus.

MD5 checksum: `r tools::md5sum(ref_aln_archive)`

--------------------------

`r path_file(taxdump_zip_file)`: Compressed file archive including NCBI
taxonomic data downloaded from https://ftp.ncbi.nlm.nih.gov/pub/taxonomy/taxdmp.zip
on `r file.info(taxdump_zip_file)$mtime`.

MD5 checksum: `r tools::md5sum(taxdump_zip_file)`

--------------------------

CHANGE LOG

2024-10-30

- Update to GenBank release 261

2024-03-14

- Update to GenBank release 259

2023-12-26

- Add taxdmp.zip

2023-12-20

- Update to GenBank release 258

2023-08-16

- Update to GenBank release 256

2023-02-16

- Update to GenBank release 253

2022-11-22

- Update to GenBank release 252

2022-09-10

- Update to GenBank release 251

2022-09-06

- Load GenBank version from targets cache
- Update references
- Add GenBank README
- Update description of restez_sql_db.tar.gz to reflect restez > v2.0.0

2022-06-23

- Add fern GenBank database

2022-05-27

- Move csv and plain text files from FigShare to Github (https://github.com/fernphy/ftol).

2022-04-04

- Update DOIs.

2022-03-31

- Generate this README file.

--------------------------

REFERENCES
