---
bibliography: references.bib
output: 
  md_document:
    pandoc_args: ["--columns=80"]
editor_options:
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
  echo = FALSE, message = FALSE,
  warning = FALSE, results = "hide", cache = FALSE)

# This is meant to be run within a drake plan.
# If editing interactively (such as in RStudio)
# un-comment the lines below to load packages and the drake cache
# 
# source(here::here("R/packages.R"))
# plastid_cache <- new_cache(here::here("plastid_cache"))

# Load objects build during analysis from drake cache
loadd(
  list = c(
    "plastid_tree_dated",
    "plastid_calibration_dates",
    "plastome_outgroups",
    "plastid_alignment",
    "plastid_genes_aligned_trimmed_renamed",
    "resolved_names_all",
    "plastid_acc_data",
    "plastid_gene_part_data"),
  cache = plastid_cache)

```

```{r stats}
# Calculate various statistics for reporting

# Combined alignment size
# - number of species
n_concat_species <- plastid_alignment %>% nrow() %>% scales::number(big.mark = ",")
# - number of base pairs
n_concat_bases <- plastid_alignment %>% ncol() %>% scales::number(big.mark = ",")
# - percent missing data
base_freqs <- ape::base.freq(plastid_alignment, all = TRUE)
p_concat_missing <- sum(base_freqs[c("n", "-", "?")]) %>% scales::percent(accuracy = 0.01)

# Number of species in tree
n_tip <- ape::Ntip(plastid_tree_dated) %>% scales::number(big.mark = ",")

# Some of the outgroup species names changed from `plastome_outgroups` data
# to the final tree:
# Physcomitrella patens -> Physcomitrium patens
# Anthoceros formosae -> Anthoceros angustus
# Fix these by joining on accession number
plastome_outgroups_correct_names <- inner_join(
  select(plastome_outgroups, accession), 
  resolved_names_all, by = "accession") %>%
  transmute(
    species,
    status = "outgroup")

# Make table of tips in tree with ingroup/outgroup status
tree_groups <-
tibble(species = plastid_tree_dated$tip.label) %>%
  mutate(species = str_replace_all(species, "_", " ")) %>%
  left_join(plastome_outgroups_correct_names, by = "species") %>%
  mutate(status = replace_na(status, "ingroup"))

n_ingroup_tips <- tree_groups %>% filter(status == "ingroup") %>% pull(species) %>% n_distinct() %>% scales::number(big.mark = ",")
n_outgroup_tips <- tree_groups %>% filter(status == "outgroup") %>% pull(species) %>% n_distinct() %>% scales::number(big.mark = ",")

```

# Fern Tree of Life (FTOL) Data Release v0.0.1 README

This README.md file was generated on `r format(Sys.time(), '%d %B, %Y')`

The goal of the FTOL project is to generate a maximally sampled phylogenetic tree for all extant fern species. It will be continuously updated as new data become available. Each release of the tree and associated metadata has a version number available on the [project github repository](https://github.com/joelnitta/ftol). The current data release includes GenBank accessions with allowed dates from 1980-01-01 to `r str_replace_all(date_cutoff, "\\/", "-")`.

## Files

Files in this data release include:

- `ftol_plastid_accs.csv`: Table of GenBank accessions used for phylogenetic analysis. Columns include `sci_name` (scientific name including author after taxonomic standardization), `species` (same as `sci_name` but without author), a series of columns named after genes (e.g., `atpA`, `rbcL`), and one column named `full_plastome`. The columns named after genes and `full_plastome` each contain the GenBank accession for that gene or full plastome. `r nrow(plastid_acc_data) %>% scales::number(big.mark = ",")` rows. MD5 hash: `r tools::md5sum(here::here("results/releases/ftol_plastid_accs.csv"))`

- `ftol_plastid_concat.fasta`: Concatenated alignment of plastid genes used for phylogenetic analysis. `r n_concat_species` species x `r n_concat_bases` base pairs, including `r length(plastid_genes_aligned_trimmed_renamed)` genes. `r p_concat_missing` missing data. FASTA format. MD5 hash: `r tools::md5sum(here::here("results/releases/ftol_plastid_concat.fasta"))`

- `ftol_plastid_parts.csv`: Table of gene partitions in concatenated alignment. Columns include `gene` (gene name), `start` (gene start position, inclusive), `end` (gene end position, inclusive). `r nrow(plastid_gene_part_data) %>% scales::number(big.mark = ",")` rows. MD5 hash: `r tools::md5sum(here::here("results/releases/ftol_plastid_parts.csv"))`

- `ftol_plastid.tre`: Fern tree of life (FTOL) based on plastid genes, not dated. `r n_tip` species (one tip per species), including `r n_ingroup_tips` ingroup and `r n_outgroup_tips` outgroup taxa. Consensus maximum-likelihood tree. Branch lengths in units of expected change per site. Values at nodes are support values as measured with ultra-fast bootstrap approximation. Newick format. MD5 hash: `r tools::md5sum(here::here("results/releases/ftol_plastid.tre"))`

- `ftol_plastid_dated.tre`: Fern tree of life (FTOL) with divergence times estimated on `ftol_plastid.tre` using fossil calibration points. Branch lengths in millions of years before present. Newick format. MD5 hash: `r tools::md5sum(here::here("results/releases/ftol_plastid_dated.tre"))`

- `plastid_cache.tar.gz`: Zipped file containing the [drake](https://docs.ropensci.org/drake/) cache with all intermediate output (R data files). MD5 hash: FIXME <!-- this can't be computed from within drake, so edit README.md manually -->

## Methods Summary

The approach leverages the fact that only a handful of genes have been intensively sequenced in ferns, and account for the vast majority of sequences on GenBank: *rbcL*, *atpA*, *atpB*, and *rps4*. These four genes (ca. 5 kb total) provide maximum phylogenetic breadth, but lack the resolving power necessary to infer a robustly supported tree across all ferns. Therefore, a set of single-copy genes are also sampled from fern plastomes, and combined with the Sanger sequences to create a mostly empty supermatrix. The plastome data provide a well-supported backbone, which is filled in by the much more diverse Sanger sequences.

The workflow is implemented in R v`r paste0(R.version$major, ".", R.version$minor)` [@RCoreTeam2020] with the Drake package v`r packageVersion("drake")` [@MichaelLandau2018]. After downloading plastid genes and genomes, taxonomic name resolution (detection of synonyms) is carried out using the [Catalog of Life (COL)](https://www.catalogueoflife.org/) as a taxonomic standard. Next, rogue taxa are detected by an all-by-all BLAST search [@Camacho2009]; any sequences whose top three hits match a different family are excluded. One specimen per species is selected prioritizing those that come from the same specimen with maximum combined length. Genes are then aligned individually using MAFFT [@Katoh2002], trimmed with TrimAl [@Capella-Gutierrez2009], then concatenated into a supermatrix. Only sequences originating from the same specimen are concatenated. The tree is inferred with IQTREE [@Nguyen2015], and dated with treePL [@Smith2012] using `r nrow(plastid_calibration_dates)` fossil calibration points.

All code is available at https://github.com/joelnitta/ftol.

A docker image to run the code is available at https://hub.docker.com/r/joelnitta/ftol

## References
